File: server/internal/service/auth_service.go
package service

import (
	"context"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"grgc/server/internal/model"
	"grgc/server/internal/storage"
	"grgc/server/pb"
)

// AuthService is the service for authentication
type AuthService struct {
	jwtManager  *JWTManager
	userStorage storage.UserStorage
	pb.UnimplementedAuthServiceServer
}

// NewAuthService returns a new auth server
func NewAuthService(jwtManager *JWTManager, userStorage storage.UserStorage) *AuthService {
	return &AuthService{
		jwtManager:  jwtManager,
		userStorage: userStorage,
	}
}

// Register creates a new user and return with credentials
func (service *AuthService) Register(ctx context.Context, req *pb.RegisterRequest) (*pb.RegisterResponse, error) {

	user := &model.User{
		Username: req.GetUsername(),
		Password: req.GetPassword(),
		Email:    req.GetEmail(),
	}

	user.HashPassword()

	err := service.userStorage.Create(user)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "cannot create user: %v", err)
	}

	accessToken, err := service.jwtManager.Generate(user.ID, service.jwtManager.accessTokenDuration)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "cannot generate access token")
	}

	refreshToken, err := service.jwtManager.Generate(user.ID, service.jwtManager.refreshTokenDuration)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "cannot generate refresh token")
	}

	res := &pb.RegisterResponse{
		AccessToken:  accessToken,
		RefreshToken: refreshToken,
		UserId:       user.ID,
	}
	return res, nil
}

// Login is a call to login user
func (service *AuthService) Login(ctx context.Context, req *pb.LoginRequest) (*pb.LoginResponse, error) {

	user, err := service.userStorage.GetByUsername(req.GetUsername())
	if err != nil {
		return nil, status.Errorf(codes.Internal, "cannot find user: %v", err)
	}

	if user == nil || !user.CheckPassword(req.GetPassword()) {
		return nil, status.Errorf(codes.NotFound, "incorrect username/password")
	}

	accessToken, err := service.jwtManager.Generate(user.ID, service.jwtManager.accessTokenDuration)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "cannot generate access token")
	}

	refreshToken, err := service.jwtManager.Generate(user.ID, service.jwtManager.refreshTokenDuration)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "cannot generate refresh token")
	}

	res := &pb.LoginResponse{
		AccessToken:  accessToken,
		RefreshToken: refreshToken,
		UserId:       user.ID,
	}
	return res, nil
}

// Refresh is a call to refresh user access token
func (service *AuthService) Refresh(ctx context.Context, req *pb.RefreshRequest) (*pb.RefreshResponse, error) {
	userClaims, err := service.jwtManager.Verify(req.RefreshToken)
	if err != nil {
		return nil, status.Errorf(codes.Unauthenticated, "refresh token is invalid: %v", err)
	}

	if userClaims.UserId == 0 {
		return nil, status.Errorf(codes.Unauthenticated, "refresh token is invalid: missing claims")
	}

	accessToken, err := service.jwtManager.Generate(userClaims.UserId, service.jwtManager.accessTokenDuration)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "cannot generate access token")
	}

	refreshToken, err := service.jwtManager.Generate(userClaims.UserId, service.jwtManager.refreshTokenDuration)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "cannot generate refresh token")
	}

	res := &pb.RefreshResponse{
		AccessToken:  accessToken,
		RefreshToken: refreshToken,
		UserId:       userClaims.UserId,
	}
	return res, nil
}

File: client/src/services/AuthService.js
import { LoginRequest, RegisterRequest, RefreshRequest, AuthServiceClient } from "../pb/auth_service_grpc_web_pb"
import { getRefreshToken } from "./utils/LocalStorageUtils";
import { getHostName } from "./utils/ServicesUtils";

let client = new AuthServiceClient(
    getHostName(),
    null,
    null);

const AuthService = {
    login(username, password, cb) {

        let loginRequest = new LoginRequest();
        loginRequest.setUsername(username);
        loginRequest.setPassword(password);

        client.login(loginRequest, null, (error, response) => {
            if (error) {
                cb(error, null)
            } else {
                cb(error, {
                    accessToken: response.getAccessToken(),
                    refreshToken: response.getRefreshToken(),
                    id: response.getUserId(),
                })
            }
        });
    },

    register(username, password, email, cb) {

        let registerRequest = new RegisterRequest();
        registerRequest.setUsername(username);
        registerRequest.setPassword(password);
        registerRequest.setEmail(email);

        client.register(registerRequest, null, (error, response) => {
            if (error) {
                cb(error, null)
            } else {
                cb(error, {
                    accessToken: response.getAccessToken(),
                    refreshToken: response.getRefreshToken(),
                    id: response.getUserId(),
                })
            }
        });
    },

    refresh(cb) {

        let refreshRequest = new RefreshRequest();
        refreshRequest.setRefreshToken(getRefreshToken());

        client.refresh(refreshRequest, null, (error, response) => {
            if (error) {
                cb(error, null)
            } else {
                cb(error, {
                    accessToken: response.getAccessToken(),
                    refreshToken: response.getRefreshToken(),
                    id: response.getUserId(),
                })
            }
        });
    },

    logout(cb) {
        localStorage.removeItem('user')
        cb(null)
    }
};

export default AuthService;
